Sources/DistributedSystem/DistributedSystem.swift

```diff
@@ -959,31 +1187,66 @@ public class DistributedSystem: DistributedActorSystem, @unchecked Sendable {
-                let queueSize = ManagedAtomic<UInt64>(0)
-                self.actors[targetID] = .clientForRemoteService(channel, continuation, queueSize)
-                Task { await self.streamTask(stream, actor, channel, queueSize) }
-                return (continuation, queueSize)
-            case let .serviceForRemoteClient(_, continuation, queueSize):
-                return (continuation, queueSize)
-            case let .clientForRemoteService(_, continuation, queueSize):
-                return (continuation, queueSize)
+                let queueState = ManagedAtomic<UInt64>(0)
+                self.actors[targetID] = .clientForRemoteService(channel, continuation, queueState)
+                Task { await self.streamTask(stream, actor, channel, queueState) }
+                return (continuation, queueState)
+            case let .serviceForRemoteClient(_, continuation, queueState):
+                return (continuation, queueState)
+            case let .clientForRemoteService(_, continuation, queueState):
+                return (continuation, queueState)
             default:
                 logger.error("\(channel.remoteAddressDescription): invalid actor state \(actorInfo)")
                 return nil
             }
         }
 
         if let res {
-            let sizeBits = (UInt64.bitWidth - 8)
+            let sizeBits = Self.endpointQueueSizeBits
             let sizeMask = ((UInt64(1) << sizeBits) - 1)
-            let queueSize = res.queueSize.wrappingIncrementThenLoad(by: envelope.size, ordering: .relaxed)
-            if (queueSize & sizeMask) >= (Self.endpointQueueWarningSize << (queueSize >> sizeBits)) {
-                // The warning threshold multiplied by 2 each time is breached,
-                // so we will have warnings for 1, 2, 4, 8, etc megabytes
-                logger.warning("Input queue size for \(envelope.targetID) reached \(queueSize & sizeMask) bytes")
-                res.queueSize.wrappingIncrement(by: (UInt64(1) << sizeBits), ordering: .relaxed)
+            var oldState = res.queueState.load(ordering: .relaxed)
+            while true {
+                var newState = (oldState + envelope.size)
+                let oldSize = (oldState & sizeMask)
+                let newSize = (newState & sizeMask)
+                let warningSize = (Self.endpointQueueWarningSize << ((oldState >> sizeBits) & 0x7F))
+                if (oldSize < warningSize) && (newSize >= warningSize) {
+                    newState += (UInt64(1) << sizeBits)
+                }
+                if (oldSize < Self.endpointQueueHighWatermark) && (newSize >= Self.endpointQueueHighWatermark) && ((oldState & Self.endpointQueueSuspendIndicator) == 0) {
+                    newState |= Self.endpointQueueSuspendIndicator
+                }
+                let (exchanged, original) = res.queueState.compareExchange(expected: oldState, desired: newState, ordering: .releasing)
```

Comment:
if I understand correctly, we only operate on atomic without changing other memory, so ordering can be `.relaxed`

Reply:
Fixed.